### PR TITLE
lightning: finer granularity for tidb backend finished bytes

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

I still use the old Lightning.Status to get the finished bytes and total bytes. For TiDB backend, finished bytes will be counted after write to TiDB, and also will recover from checkpoints

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by lance6716** [comment]
/cc @D3Hunter @okJiang 

I'm working on it

File: br/pkg/lightning/restore/table_restore.go [hunk 1]
```diff
@@ -475,6 +480,9 @@ func (tr *TableRestore) restoreEngine(
 
 	// Restore table data
 	for chunkIndex, chunk := range cp.Chunks {
+		if rc.status != nil && rc.status.backend == config.BackendTiDB {
+			rc.status.FinishedFileSize.Add(chunk.Chunk.Offset)
```
Comment (D3Hunter) [inline br/pkg/lightning/restore/table_restore.go:484]: should it be `chunk.Chunk.Offset - chunk.Key.Offset` to count previous imported data

File: br/pkg/lightning/restore/table_restore.go [hunk 2]
```diff
@@ -406,6 +406,11 @@ func (tr *TableRestore) restoreEngine(
 		if err != nil {
 			return closedEngine, errors.Trace(err)
 		}
+		if rc.status != nil && rc.status.backend == config.BackendTiDB {
+			for _, chunk := range cp.Chunks {
+				rc.status.FinishedFileSize.Add(chunk.Chunk.EndOffset)
```
Comment (okJiang) [inline br/pkg/lightning/restore/table_restore.go:411]: I am not sure TiDB backend would run here(`cp.Status >= checkpoints.CheckpointStatusAllWritten`)😵‍💫?

[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/restore/table_restore.go:411]: I'm not familiar with strict format, can you give some reference to the behaviour?

I think you want to keep a property that the value of `FinishedFileSize.Add` should be same no matter it's processed without failure or recovered from checkpoint?

File: br/pkg/lightning/restore/table_restore.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/restore/table_restore.go:484]: here I want to recover FinishedFileSize from checkpoint, the restore process of current run is not started

File: br/pkg/lightning/restore/table_restore.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/restore/table_restore.go:411]: I found there's a saving time for tidb backend

https://github.com/pingcap/tidb/blob/65585a5b3cd1dc13c75a8f2aa1b7d87a1303434e/br/pkg/lightning/restore/table_restore.go#L622-L623

[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/restore/table_restore.go:411]: OK I guess the strict mode will split a file into many chunks, the split position is written in Key.Ofset

**Review by D3Hunter — APPROVED** [review]

**Review by okJiang — APPROVED** [review]

**Pull request merged**